### PR TITLE
Make perplexity evaluation general

### PR DESCRIPTION
**Pull request by pashminacameron**

### CODE REVIEW

File: experiments/run_finetuning.py [hunk 1]
```diff
@@ -302,11 +302,11 @@ def main() -> None:
...
 
     model = model_adapter.model
-    model = get_peft_model(model, lora_config)
-    model.print_trainable_parameters()
+    lora_model = get_peft_model(model, lora_config)
+    lora_model.print_trainable_parameters()
 
     # create optimizer and scheduler
-    optimizer, lr_scheduler = get_optimizer_and_scheduler(model, finetune_ds["train"], args)
+    optimizer, lr_scheduler = get_optimizer_and_scheduler(lora_model, finetune_ds["train"], args)
```
Comment (pashminacameron) [inline experiments/run_finetuning.py:309]: This line is the change that triggered the refactoring. We want to calculate perplexity on fine-tuned model which doesn't have an adapter.

File: experiments/run_finetuning.py [hunk 2]
```diff
@@ -348,15 +348,15 @@ def main() -> None:
...
+        merged_model = lora_model.merge_and_unload()
 
         torch.save(merged_model.state_dict(), model_file)
         logging.info(f"Saved sliced and finetuned model to {args.save_dir}")
 
     utils.cleanup_memory()
 
     # compute perplexity after finetuning
-    dataset_ppl = gpu_utils.evaluate_ppl(model, ppl_eval_loader)
+    dataset_ppl = gpu_utils.evaluate_ppl(lora_model, model.config.pad__token_id, ppl_eval_loader)
```
Comment (nailimixaM) [inline experiments/run_finetuning.py:359]: typo: double underscore in `pad__token_id`

File: src/slicegpt/gpu_utils.py [hunk 3]
```diff
@@ -25,17 +25,17 @@ def evaluate_ppl(model_adapter: ModelAdapter, testloader: DataLoader[dict[str, t
 
     start_time = time.time()
 
-    model_adapter.model.eval()
+    model.eval()
 
-    loss_fn = torch.nn.CrossEntropyLoss(reduction="none", ignore_index=model_adapter.model.config.pad_token_id)
```
Comment (nailimixaM) [inline src/slicegpt/gpu_utils.py:30]: Suggestion: the `pad_token_id` is from `model` so can we exclude it from the method signature and do `loss_fn = torch.nn.CrossEntropyLoss(reduction="none", ignore_index=model.config.pad_token_id)` ? (with appropriate logic to do `ignore_index=None` if `model.config` doesn't have a `pad_token_id`)

**Review by nailimixaM — APPROVED** [review]
Thanks for these changes, very satisfying to see the ppl calc be cleaner!

Approving assuming typo gets fixed.

File: src/slicegpt/gpu_utils.py (hunk 3 above)
Comment (LianaMikael) [reply src/slicegpt/gpu_utils.py:30]: Ignoring padding tokens also means that it will be ignoring bos and eos tokens too since there are all the same now. I suggest removing this and explicitly providing the mask when computing the loss.

Comment (pashminacameron) [reply src/slicegpt/gpu_utils.py:30]: In trying to do this, I discovered two things. config.pad_token_id will be either an index or None. Also, for the lora_model, we will need to pass in the ignore_indices separately because lora_model object does not have this information, so perhaps we need an overload.

**Review by LianaMikael — APPROVED** [review]

File: src/slicegpt/gpu_utils.py (hunk 3 above)
Comment (pashminacameron) [reply src/slicegpt/gpu_utils.py:30]: Updated to account for pad_token being None. Used Pavel's tests to verify no change in results. Tests pass, hence merging now.
